node/node.go

```diff
@@ -786,13 +786,16 @@ func (node *AlgorandFullNode) checkForParticipationKeys() {
 	defer node.monitoringRoutinesWaitGroup.Done()
 	ticker := time.NewTicker(node.config.ParticipationKeysRefreshInterval)
 	for {
+		node.mu.RLock()
+		done := node.ctx.Done()
+		node.mu.RUnlock()
 		select {
 		case <-ticker.C:
 			err := node.loadParticipationKeys()
 			if err != nil {
 				node.log.Errorf("Could not refresh participation keys: %v", err)
 			}
-		case <-node.ctx.Done():
+		case <-done:
```

Comment:
which test did you see this race come from? (and what was the last access to node.ctx recorded by the race detector?)

there are two places where this thread is launched (AlgorandFullNode.Start and SetCatchpointCatchupMode) and I was curious if it was the catchup mode one.

Reply:
I found https://app.circleci.com/pipelines/github/algorand/go-algorand/6331/workflows/ff6442e3-7fc2-4654-b109-e8e14a0013d7/jobs/108678 OK I see it is catchpointCatchupTest.exp.

Reply:
OK! LGTM.

Reply:
That way these threads are bound to be cancelled by the node context that launched them, not whatever happens to be the current value of node.ctx. You wouldn't need the lock this way, either.

Reply:
All the usage I've seen of Context.Done() is
`case <-ctx.Done():`
... so I'm suspicious of this usage of getting the chan outside of the switch. Maybe it's okay? Maybe I'm not sure `ctx.Done()` always needs to return the same object? cce's point that this should be stable and not need a lock also feels right to me.

Reply:
Let me take another look...Acquiring the lock outside of the for loop seems fine to me. 

Ultimately, from what I understand we _have_ to have the lock because in https://github.com/algorand/go-algorand/blob/master/node/node.go#L1232 we're writing the context variable before the cancel function has been called which means `ctx.Done()` may never be executed since it could retrieve the new channel before the cancel function invokes that branch of the select statement.

Reply:
in SetCatchpointCatchupMode, when switching modes, it calls the cancelFunc for the current node.ctx, then it creates a new ctx+cancelFunc pair and sets node.ctx. So unless you have a way to know that the 3 goroutines that are polling on node.ctx.Done() after calling the old cancelFunc, you will probably set off the race detector sometimes..?

The context.Done interface docs say "successive calls to Done return the same value." so I think it is safe to just grab ctx.Done() when each of these 3 goroutines start and keep polling on them.

Reply:
>so I'm suspicious of this usage of getting the chan outside of the switch. Maybe it's okay? Maybe I'm not sure ctx.Done() always needs to return the same object? cce's point that this should be stable and not need a lock also feels right to me.

Looks like Brian and right the [Done()](https://github.com/golang/go/blob/master/src/context/context.go#L358) is way more non-trivial than I expected.

The problem with `SetCatchpointCatchupMode` is it re-assigns `node.ctx` under a lock, and then threads use it without locking. I agree, my proposed fix that Eric implemented does not look correct. The plan was is to go with this fix to unblock others and then re-shuffle `SetCatchpointCatchupMode`  to stop goroutines first, assign, run after - more like `node.Start()` and `node.Stop()`

Reply:
> in SetCatchpointCatchupMode, when switching modes, it calls the cancelFunc for the current node.ctx, then it creates a new ctx+cancelFunc pair and sets node.ctx.

Not sure if I'm misunderstanding, but it seems like the exact opposite of that. It [writes a new context](https://github.com/algorand/go-algorand/blob/master/node/node.go#L1232) then [calls the cancel func on the old one](https://github.com/algorand/go-algorand/blob/master/node/node.go#L1235). So if we keep calling `node.ctx.Done()` as we are currently doing, we could end up with the following:

1. Node started, Context<sub>original</sub> initialized, monitoring threads started and polling `<-node.ctx.Done()` receiving on the channel.
2. Catchpoint mode invoked, `node.ctx` overwritten with a new context, Context<sub>new</sub>.
3. Monitoring threads are now receiving the new channel when calling `node.ctx.Done()`
4. Catchpoint mode invokes the cancel func of Context<sub>original</sub>.
5. Monitoring threads are not cancelled because they are monitoring Context<sub>new</sub>.
6. Catchpoint mode finishes and starts a second set of monitoring threads.

> so I think it is safe to just grab ctx.Done() when each of these 3 goroutines start and keep polling on them.
I agree that this probably works as well.

Reply:
>The context.Done interface docs say "successive calls to Done return the same value." 

Right, I looked deeper into `c.done` usage and it is the same, and `Done()` just loads from non-changed `c.done`.

Reply:
I tried something similar to this last night and it was timing out on https://github.com/algorand/go-algorand/blob/master/test/e2e-go/features/catchup/catchpointCatchup_test.go#L224 
Not sure if there are minor differences w/ your changes that will behave differently.

The version I've just pushed tries to address the core problem which is that the node mutex has really high lock contention because it's locking so many different things. So I've split out this into a context-specific mutex. Haven't run into test failures locally with this. Let me know what you think.

Reply:
Hrmm OK well I was hoping if you needed a lock, you would only have to grab the lock once, when the goroutine starts. I think it is a bug to grab the done channel more than once that could lead to orphaned goroutines. Was curious so I wrote up https://github.com/algorand/go-algorand/pull/3955/files to have CI run it...

Reply:
Looks like it's indeed passing. Let's discuss this afternoon pros and cons of each approach and figure out what we want.

Reply:
follow-up — talked with @Eric-Warehime and just explicitly passing the ctx.Done() channel to each goroutine as an argument seemed like the clearest way to avoid worrying about unnecessary synchronization https://github.com/algorand/go-algorand/pull/3955/commits/7878d412188889245a611cf185cc4b4e4ed0274c